jdaviz/configs/default/plugins/line_lists/line_lists.py

```diff
@@ -88,29 +106,183 @@ def _on_viewer_data_changed(self, msg=None):
         # However, if the msg is not none, check to make sure that it's the
-        # viewer we care about.
-        if msg is not None and msg.viewer_id != self._viewer_id:
+        # viewer we care about and that the message contains the data label.
+        if msg is None or msg.viewer_id != self._viewer_id or msg.data is None:
             return
 
+        label = msg.data.label
         try:
-            viewer_data = self.app.get_viewer('spectrum-viewer').data()
+            viewer_data = self.app.get_data_from_viewer('spectrum-viewer').get(label)
         except TypeError:
             warn_message = SnackbarMessage("Line list plugin could not retrieve data from viewer",
                                            sender=self, color="error")
             self.hub.broadcast(warn_message)
             return
 
         # If no data is currently plotted, don't attempt to update
-        if viewer_data is None or len(viewer_data) == 0:
+        if viewer_data is None:
+            return
+
+        self._units["x"] = str(viewer_data.spectral_axis.unit)
+        self._units["y"] = str(viewer_data.flux.unit)
+
+        self._bounds["min"] = viewer_data.spectral_axis[0]
+        self._bounds["max"] = viewer_data.spectral_axis[-1]
+
+        # set redshift slider to redshift stored in Spectrum1D object
+        self.rs_redshift = viewer_data.redshift.value
+        self._auto_slider_range()  # will also trigger _auto_slider_step
+
+    def _parse_redshift_msg(self, msg):
+        '''
+        Handle incoming redshift messages from the app hub. Generally these
+        will be created by Specviz helper methods.
+        '''
+        if msg.sender == self:
+            return
+
+        param = msg.param
+
+        if param == "rs_slider_range":
+            if msg.value == 'auto':
+                # observer will handle setting rs_slider_range
+                self.rs_slider_range_auto = True
+            else:
+                self.rs_slider_range_auto = False
+                self.rs_slider_half_range = float(msg.value)/2
+        elif param == "rs_slider_step":
+            if msg.value == 'auto':
+                # observer will handle setting rs_slider_step
+                self.rs_slider_step_auto = True
+            else:
+                self.rs_slider_step_auto = False
+                self.rs_slider_step = float(msg.value)
+        elif param == "redshift":
+            # NOTE: this should trigger the observe to update rs_rv, line positions, and
+            # update self._global_redshift
+            self.rs_redshift = float(msg.value)
+        elif param == 'rv':
+            # NOTE: this should trigger the observe to update rs_redshift, line positions, and
+            # update self._global_redshift
+            self.rs_rv = float(msg.value)
+        else:
+            raise NotImplementedError(f"RedshiftMessage with param {param} not implemented.")
+
+    def _velocity_to_redshift(self, velocity):
```

Comment:
See astropy/astropy#12709 . Maybe after `astropy` 5.1 is out, we can use that instead...

Reply:
sure!  Although dropping unit conversion did help with some overhead since this is called on every slider update, so we will need to consider any time cost (which we'll need to do anyways if we want to support reading/setting RVs in other units)

Reply:
Re: performance -- Is it possible to only convert the value when mouse release is detected on the slider (i.e., user stops sliding)?

Reply:
yes, that's an option, but I think its nice to see the RV updating live.  (Note: we're no longer worried about performance as-is and have a working version outside this PR which will drastically improve lagginess).

Reply:
If this conversion is not a true bottleneck for performance, might be better to use what is in astropy when that is available. That way, you know that functionality is vigorously tested upstream and it would natively support Quantity, so you can remove the comment about "if supporting non-km/s units in the future". Something to think about.